keen/tests/saved_query_tests.py

```diff
@@ -93,33 +97,106 @@ def test_create_saved_query(self):
 
     @responses.activate
     def test_update_saved_query(self):
-        saved_queries_response = {
+        unacceptable_attr = "run_information"
+        metadata_attr_name = "metadata"
+
+        original_query = {
             "query_name": "saved-query-name",
             "refresh_rate": 14400,
             "query": {
                 "analysis_type": "average",
                 "event_collection": "TheCollection",
                 "target_property": "TheProperty",
                 "timeframe": "this_2_weeks"
-                }
+            },
+            metadata_attr_name: { "foo": "bar" },
+            unacceptable_attr: { "foo": "bar" }
         }
+
         url = "{0}/{1}/projects/{2}/queries/saved/saved-query-name".format(
             self.client.api.base_url,
             self.client.api.api_version,
             self.exp_project_id
         )
 
         responses.add(
-            responses.GET, url, status=200, json=saved_queries_response
+            responses.GET, url, status=200, json=original_query
         )
 
-        responses.add(
-            responses.PUT, url, status=200, json=saved_queries_response
+        updated_query = { "query": {} }
+        new_analysis_type = "sum"
+        updated_query["query"]["analysis_type"] = new_analysis_type
+
+        def request_callback(request):
+            payload = json.loads(request.body)
+
+            # Ensure update() round-trips some necessary things like "metadata"
+            self.assertEqual(payload[metadata_attr_name], original_query[metadata_attr_name])
+
+            # Ensure update() doesn't pass unacceptable attributes
+            self.assertNotIn(unacceptable_attr, payload)
+
+            # Ensure update() merges deep updates
+            self.assertEqual(payload["query"]["analysis_type"], new_analysis_type)
+            payload["query"]["analysis_type"] = "average"
+            payload[unacceptable_attr] = original_query[unacceptable_attr]
+            self.assertEqual(payload, original_query)
+
+            headers = {}
+            return (200, headers, json.dumps(updated_query))
+
+        responses.add_callback(
+            responses.PUT,
+            url,
+            callback=request_callback,
+            content_type='application/json',
         )
 
-        saved_query = self.client.saved_queries.update("saved-query-name", saved_queries_response)
+        saved_query = self.client.saved_queries.update("saved-query-name", updated_query)
 
-        self.assertEquals(saved_query, saved_queries_response)
+        self.assertEqual(saved_query, updated_query)
```

Comment:
I could return a copy of the `payload` before altering it for the `self.assertEqual(payload, original_query)` check, but in reality, all the tests in this file and a lot of the Python tests are pretty strictly unit tests in that they don't much care what the other modules are doing. So as long as the thing that gets returned from the `PUT` request isn't altered before it gets returned to me, that's what's important. You're right though, I *could* do a deep copy of `payload` and alter that for the assertion, and make `payload` a var local to the test and not the callback, but I don't think it changes much w.r.t. what's being asserted in the last `assertEqual()` call. Really I think that last assert is less important and more in keeping with the rest of the tests, and just to make sure we don't add some intermediary code later that alters that return value.